### PR TITLE
Improvements and test coverage for `line.py`

### PR DESCRIPTION
**Pull request by Aufinal**
Changes to `line.py` itself:
 * due to the discussion in the documentation of `line_graph`, line graphs are generated with no loops.
    It makes therefore sense to exclude graphs with loops from `inverse_line_graphs`
* we can always pop u from `partitioned_vertices` since the cell `new_cell` contains all the neighbors of u,
  and thus `deg_u = 0` after line 403.
* we only ever use the length of ac_edges and bc_edges
* the only way to have `len(triangle_nodes) != s+2` is if `G` has self-loops (otherwise each odd triangle contributes
  exactly one new vertex to `triangle_nodes`).

Test changes:
* adds coverage for working inverse line graph examples
* adds a bunch of negative examples (the whole list of forbidden subgraphs of L.W.Beineke), which also covers all possible branches
* adds some very basic tests for private internal functions

### CODE REVIEW

**Comment by Aufinal** [comment]
Note : I had missed #6204 when working on that (sorry...), but I do have a test that hits [those lines](https://github.com/networkx/networkx/pull/6215/files#diff-4e6352234d8b3f17771dd2ea80891796b749790a73596994000dc4bea591045fR293-R295), so they are not unreachable as claimed.

**Comment by rossbar** [comment]
> I had missed https://github.com/networkx/networkx/pull/6204 when working on that (sorry...), but I do have a test that hits [those lines](https://github.com/networkx/networkx/pull/6215/files#diff-4e6352234d8b3f17771dd2ea80891796b749790a73596994000dc4bea591045fR293-R295), so they are not unreachable as claimed.

No worries @Aufinal , thanks for pointing this out. @Mjh9122 want to take a look at this one?

**Comment by Aufinal** [comment]
I'm not sure that matching on the precise error messages is meaningful: they depend on a lot of hidden details, such as the order of edges in G and even the order of vertices in edges, so a lot of unrelated API changes could cause the tests to fail. That's also why there are a lot of redundant negative examples, to try and make sure the coverage will stay OK no matter what.

With that in mind, I've replaced some tests with one less dependent on the precise API. I welcome any suggestions about the self-loop behavior, but I don't know the best way to proceed either...

**Comment by Aufinal** [comment]
Changed the behavior for self-loops (and the associated tests) to a simple warning. I am making a copy of G to avoid removing edges in-place, is that the correct to deal with such cases @rossbar ?

**Comment by rossbar** [comment]
> The only question I have is about whether warnings are the right way to handle graphs with selfloops.

Agreed - let's avoid warnings. `inverse_line_graph` currently raises at least for some subset of cases involving self-loops:

```python
>>> G = nx.path_graph(2)
>>> L_1 = nx.inverse_line_graph(G)  # Works fine
>>> G.add_edge(0, 0)
>>> nx.inverse_line_graph(G)
Traceback (most recent call last)
   ...
NetworkXError: G is not a line graph (odd triangles do not form complete subgraph)
```

I would vote to preserve this behavior, though raising an exception message specific to the self-loop case would be an improvement!

**Comment by Aufinal** [comment]
Agreed with the sentiment of keeping the current behavior as much as possible. I've changed the selfloop behavior to raise an error.

File: networkx/generators/tests/test_line.py [hunk 1]
```diff
@@ -1,3 +1,5 @@
+from itertools import permutations
+
```
Comment (dschult) [inline networkx/generators/tests/test_line.py:2]:
```suggestion
```

**Review by dschult — COMMENTED** [review]
This is very close to being ready to merge. I have some minor suggestions below and it looks like we'll need to merge "main" into this branch.  If you don't have time let me know and I'll do it.

**Comment by Aufinal** [comment]
The suggestions are applied, and I think the merge went ok :)

**Review by dschult — APPROVED** [review]
This looks good to me!
Thanks @Aufinal !

**Review by MridulS — APPROVED** [review]
Thanks @Aufinal!

**Pull request merged**